File: app/src/main/java/com/joekickass/mondaymadness/intervaltimer/IntervalTimer.java
package com.joekickass.mondaymadness.intervaltimer;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Controller for {@link IntervalTimerView}
 *
 * Handles two types of intervals; WORK and REST. It is also possible to create a chain of intervals
 * by specifying the number of repetitions.
 */
public class IntervalTimer implements IntervalTimerView.Callback {

    private final long mWorkInMillis;

    private final long mRestInMillis;

    private final int mRepetitions;

    public interface IntervalTimerListener {
        void onWorkFinished();
        void onRestFinished();
        void onIntervalTimerFinished();
    }

    private final IntervalTimerView mView;

    private Queue<Interval> mWorkQueue;

    private Interval mCurrent;

    // TODO: Create special notifier/listener class with thread handling?
    private List<IntervalTimerListener> mListeners = new CopyOnWriteArrayList<>();

    public IntervalTimer(IntervalTimerView view, long workInMillis, long restInMillis, int repetitions) {
        mWorkInMillis = workInMillis;
        mRestInMillis = restInMillis;
        mRepetitions = repetitions;
        mView = view;
        mView.registerCallback(this);
        reset();
    }

    public void start() {
        mView.start();
    }

    public void pause() {
        mView.pause();
    }

    public boolean isRunning() {
        return mView.isRunning();
    }

    public void reset() {
        mWorkQueue = generateWorkQueue(mWorkInMillis, mRestInMillis, mRepetitions);
        mCurrent = mWorkQueue.poll();
        mView.init(mCurrent.time);
    }

    @Override
    public void onIntervalFinished() {

        // Notify listeners
        for (IntervalTimerListener listener : mListeners) {
            switch (mCurrent.type) {
                case WORK:
                    listener.onWorkFinished();
                    break;
                case REST:
                    listener.onRestFinished();
                    break;
            }
        }

        // Start new if there are any intervals left
        if (!mWorkQueue.isEmpty()) {
            mCurrent = mWorkQueue.poll();
            mView.init(mCurrent.time);
            mView.start();
        } else {
            for (IntervalTimerListener listener : mListeners) {
                listener.onIntervalTimerFinished();
            }
        }
    }

    public void addListener(IntervalTimerListener listener) {
        mListeners.add(listener);
    }

    public void removeListener(IntervalTimerListener listener) {
        mListeners.remove(listener);
    }

    private Queue<Interval> generateWorkQueue(long work, long rest, int repetitions) {
        Queue<Interval> ret = new LinkedList<>();
        for (int i = 1; i <= repetitions; ++i) {
            ret.add(new Interval(Interval.IntervalType.WORK, work));
            ret.add(new Interval(Interval.IntervalType.REST, rest));
        }
        return ret;
    }

    private static class Interval {
        private enum IntervalType { WORK, REST; }
        Interval(IntervalType type, long time) {
            this.type = type;
            this.time = time;
        }
        IntervalType type;
        long time;
    }
}

File: app/src/main/kotlin/com/joekickass/mondaymadness/intervaltimer/IntervalTimer.kt
package com.joekickass.mondaymadness.intervaltimer

import com.joekickass.mondaymadness.model.Timer
import com.joekickass.mondaymadness.model.Workout

/**
 * Controller for [IntervalTimerView]
 */
class IntervalTimer(private val view: IntervalTimerView,
                    workInMillis: Long,
                    restInMillis: Long,
                    repetitions: Int) {

    interface IntervalTimerListener {
        fun onWorkFinished()
        fun onRestFinished()
        fun onIntervalTimerFinished()
    }

    private val workout = Workout(workInMillis, restInMillis, repetitions)

    private var timer: Timer = Timer(0)

    // TODO: Create special notifier/listener class with thread handling?
    private val mListeners = mutableListOf<IntervalTimerListener>()

    init {
        timer = Timer(workout.time)
        view.init(timer, { intervalFinished() })
    }

    fun start() {
        view.start()
    }

    fun pause() {
        view.pause()
    }

    val isRunning: Boolean
        get() = timer.isRunning

    fun addListener(listener: IntervalTimerListener) {
        mListeners.add(listener)
    }

    fun removeListener(listener: IntervalTimerListener) {
        mListeners.remove(listener)
    }

    private fun intervalFinished() {

        // Notify listeners
        for (listener in mListeners) {
            if (workout.work) listener.onWorkFinished()
            if (workout.rest) listener.onRestFinished()
        }

        // Start new if there are any intervals left
        if (workout.hasNext()) {
            workout.next()
            timer = Timer(workout.time)
            view.init(timer, { intervalFinished() })
            view.start()
        } else {
            for (listener in mListeners) {
                listener.onIntervalTimerFinished()
            }
        }
    }
}